internal/cache/file/downloader/parallel_downloads_job.go

```diff
@@ -0,0 +1,182 @@
+	cacheFile, err := cacheutil.CreateFile(job.fileSpec, os.O_TRUNC|os.O_RDWR)
+	if err != nil {
+		err = fmt.Errorf("parallelDownloadObjectAsync: error in creating cache file: %w", err)
+		job.failWhileDownloading(err)
+		return
+	}
+	defer func() {
+		err = cacheFile.Close()
+		if err != nil {
+			err = fmt.Errorf("parallelDownloadObjectAsync: error while closing cache file: %w", err)
+			job.failWhileDownloading(err)
+		}
+	}()
+
+	notifyInvalid := func() {
+		job.mu.Lock()
+		job.status.Name = Invalid
+		job.notifySubscribers()
+		job.mu.Unlock()
+	}
+
+	var start, end int64
+	end = int64(job.object.Size)
+	var parallelReadRequestSize = int64(job.fileCacheConfig.ReadRequestSizeMB) * cacheutil.MiB
+	var maxSingleStepReadSize = parallelReadRequestSize * int64(job.fileCacheConfig.DownloadParallelismPerFile)
+
+	for {
+		select {
+		case <-job.cancelCtx.Done():
+			return
+		default:
+			if start < end {
+				// Parallely download different ranges not more than maxSingleStepReadSize
+				// and not using go routines more than job.downloadParallelism
+				downloadErrGroup, downloadErrGroupCtx := errgroup.WithContext(job.cancelCtx)
+
+				var singleStepReadSize int64 = 0
+				for goRoutineIdx := 0; (goRoutineIdx < job.fileCacheConfig.DownloadParallelismPerFile) &&
+					(singleStepReadSize < maxSingleStepReadSize) && (start < end); goRoutineIdx++ {
+					rangeStart := start
+					rangeEnd := min(rangeStart+parallelReadRequestSize, end)
+
+					downloadErrGroup.Go(func() error {
+						return job.downloadRangeToFile(downloadErrGroupCtx, cacheFile, rangeStart, rangeEnd)
+					})
+
+					singleStepReadSize = singleStepReadSize + rangeEnd - rangeStart
+					start = rangeEnd
+				}
+				// If any of the go routine failed, consider the async job failed.
+				err = downloadErrGroup.Wait()
+				if err != nil {
+					if errors.Is(err, context.Canceled) {
+						notifyInvalid()
+						return
+					}
+					job.failWhileDownloading(err)
+					return
+				}
+
+				job.mu.Lock()
+				job.status.Offset = start
+				err = job.updateFileInfoCache()
+				// Notify subscribers if file cache is updated.
+				if err == nil {
+					job.notifySubscribers()
+				} else if strings.Contains(err.Error(), lru.EntryNotExistErrMsg) {
+					// Download job expects entry in file info cache for the file it is
+					// downloading. If the entry is deleted in between which is expected
+					// to happen at the time of eviction, then the job should be
+					// marked Invalid instead of Failed.
+					job.status.Name = Invalid
+					job.notifySubscribers()
+					logger.Tracef("Job:%p (%s:/%s) is no longer valid due to absense of entry in file info cache.", job, job.bucket.Name(), job.object.Name)
+					job.mu.Unlock()
+					return
+				}
+				job.mu.Unlock()
+				// Change status of job in case of error while updating file cache.
+				if err != nil {
+					job.failWhileDownloading(err)
+					return
+				}
+			} else {
+				err = job.validateCRC()
```

Comment:
why are we duplicating the code so much?

Reply:
I tried to but it's not straightforward to break down the function into logically independent blocks. 
Trying one more thing, hopefully that should (but would require additional unit testing)

Reply:
refactored. Will add unit tests in next PR for refactored methods.